// @devrev/typescript-sdk/dist/auto-generated/public-devrev-sdk.js
"use strict";
/* eslint-disable */
/* tslint:disable */
/*
 * ---------------------------------------------------------------
 * ## THIS FILE WAS GENERATED VIA SWAGGER-TYPESCRIPT-API        ##
 * ##                                                           ##
 * ## AUTHOR: acacode                                           ##
 * ## SOURCE: https://github.com/acacode/swagger-typescript-api ##
 * ---------------------------------------------------------------
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlaSelectorAppliesTo = exports.SlaEvaluationPeriod = exports.SlaAppliesTo = exports.SchemaFieldDescriptorFieldType = exports.SchemaFieldDescriptorArrayTypeBaseType = exports.QuestionAnswerStatus = exports.PartType = exports.OrgType = exports.OrgScheduleStatus = exports.OrgEnvironment = exports.MetricDefinitionStatus = exports.MetricDefinitionMetricType = exports.MetricDefinitionAppliesTo = exports.MemberType = exports.ListMode = exports.LinksDirection = exports.LinkType = exports.LinkEndpointType = exports.IssuePriority = exports.GroupType = exports.GroupMemberType = exports.GroupIngestionSource = exports.EventFetchedResult = exports.ErrorUnauthorizedType = exports.ErrorTooManyRequestsType = exports.ErrorServiceUnavailableType = exports.ErrorNotFoundType = exports.ErrorInternalServerErrorType = exports.ErrorForbiddenType = exports.ErrorConflictType = exports.ErrorBadRequestUnexpectedJsonTypeType = exports.ErrorBadRequestType = exports.DevUserJobTitle = exports.DevOrgAuthConnectionsUpdateRequestType = exports.DevOrgAuthConnectionsCreateRequestType = exports.DateTimePresetType = exports.DateFilterType = exports.ConversationsCreateRequestTypeValue = exports.BooleanExpressionType = exports.AuthTokenTokenType = exports.AuthTokenSubjectTokenType = exports.AuthTokenStatus = exports.AuthTokenRequestedTokenType = exports.AuthTokenGrantType = exports.AuthConnectionType = exports.AuthConnectionToggle = exports.AtomType = exports.ArticleType = exports.ArticleStatus = exports.AccessLevel = void 0;
exports.Api = exports.HttpClient = exports.ContentType = exports.WorkType = exports.WebhooksUpdateAction = exports.WebhookStatus = exports.WebhookEventType = exports.VistaGroupItemType = exports.VistaGroupItemState = exports.UserType = exports.UserState = exports.TimelineEntryVisibility = exports.TimelineEntryType = exports.TimelineEntryObjectType = exports.TimelineEntriesUpdateRequestType = exports.TimelineEntriesCreateRequestType = exports.TimelineCommentBodyType = exports.TicketSeverity = exports.TicketChannels = exports.SyncRunMode = exports.SyncProgressState = exports.SyncMetadataFilterSyncOutFilterStatus = exports.SyncMetadataFilterSyncInFilterStatus = exports.StageValidationOptionForUpdate = exports.StageValidationOptionForCreate = exports.SlasFilterAppliesToOperatorType = exports.SlaType = exports.SlaTrackerAppliesToType = exports.SlaSummaryStage = exports.SlaStatus = exports.SlaSelectorSeverity = exports.SlaSelectorPriority = void 0;
var AccessLevel;
(function (AccessLevel) {
    AccessLevel["External"] = "external";
    AccessLevel["Internal"] = "internal";
    AccessLevel["Private"] = "private";
    AccessLevel["Public"] = "public";
    AccessLevel["Restricted"] = "restricted";
})(AccessLevel = exports.AccessLevel || (exports.AccessLevel = {}));
/** Status of the article. */
var ArticleStatus;
(function (ArticleStatus) {
    ArticleStatus["Archived"] = "archived";
    ArticleStatus["Draft"] = "draft";
    ArticleStatus["Published"] = "published";
    ArticleStatus["ReviewNeeded"] = "review_needed";
})(ArticleStatus = exports.ArticleStatus || (exports.ArticleStatus = {}));
/** Type of the article. */
var ArticleType;
(function (ArticleType) {
    ArticleType["Article"] = "article";
    ArticleType["ContentBlock"] = "content_block";
})(ArticleType = exports.ArticleType || (exports.ArticleType = {}));
var AtomType;
(function (AtomType) {
    AtomType["Capability"] = "capability";
    AtomType["Conversation"] = "conversation";
    AtomType["DevUser"] = "dev_user";
    AtomType["Enhancement"] = "enhancement";
    AtomType["Feature"] = "feature";
    AtomType["Issue"] = "issue";
    AtomType["Product"] = "product";
    AtomType["RevOrg"] = "rev_org";
    AtomType["RevUser"] = "rev_user";
    AtomType["SysUser"] = "sys_user";
    AtomType["Tag"] = "tag";
    AtomType["Ticket"] = "ticket";
    AtomType["TimelineComment"] = "timeline_comment";
    AtomType["Webhook"] = "webhook";
})(AtomType = exports.AtomType || (exports.AtomType = {}));
var AuthConnectionToggle;
(function (AuthConnectionToggle) {
    AuthConnectionToggle["Disable"] = "disable";
    AuthConnectionToggle["Enable"] = "enable";
})(AuthConnectionToggle = exports.AuthConnectionToggle || (exports.AuthConnectionToggle = {}));
/**
 * Defines the type for the authentication connection. The configuration
 * for each authentication connection will depend on the type value.
 */
var AuthConnectionType;
(function (AuthConnectionType) {
    AuthConnectionType["GoogleApps"] = "google_apps";
    AuthConnectionType["Oidc"] = "oidc";
    AuthConnectionType["Okta"] = "okta";
    AuthConnectionType["Samlp"] = "samlp";
    AuthConnectionType["Social"] = "social";
    AuthConnectionType["Waad"] = "waad";
})(AuthConnectionType = exports.AuthConnectionType || (exports.AuthConnectionType = {}));
/** Specifies the process of obtaining a token. */
var AuthTokenGrantType;
(function (AuthTokenGrantType) {
    AuthTokenGrantType["UrnDevrevParamsOauthGrantTypeTokenIssue"] = "urn:devrev:params:oauth:grant-type:token-issue";
    AuthTokenGrantType["UrnIetfParamsOauthGrantTypeTokenExchange"] = "urn:ietf:params:oauth:grant-type:token-exchange";
})(AuthTokenGrantType = exports.AuthTokenGrantType || (exports.AuthTokenGrantType = {}));
/** The type of the requested token. */
var AuthTokenRequestedTokenType;
(function (AuthTokenRequestedTokenType) {
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeAat"] = "urn:devrev:params:oauth:token-type:aat";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeAatActAs"] = "urn:devrev:params:oauth:token-type:aat:act-as";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeAatPublic"] = "urn:devrev:params:oauth:token-type:aat:public";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeDev"] = "urn:devrev:params:oauth:token-type:dev";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeDevConnect"] = "urn:devrev:params:oauth:token-type:dev:connect";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeGat"] = "urn:devrev:params:oauth:token-type:gat";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypePat"] = "urn:devrev:params:oauth:token-type:pat";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypePatActAs"] = "urn:devrev:params:oauth:token-type:pat:act-as";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeRat"] = "urn:devrev:params:oauth:token-type:rat";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeRev"] = "urn:devrev:params:oauth:token-type:rev";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeSession"] = "urn:devrev:params:oauth:token-type:session";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeSessionDev0"] = "urn:devrev:params:oauth:token-type:session:dev0";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeSessionOnetime"] = "urn:devrev:params:oauth:token-type:session:onetime";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeSessionRevActAs"] = "urn:devrev:params:oauth:token-type:session:rev:act-as";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeSuper"] = "urn:devrev:params:oauth:token-type:super";
    AuthTokenRequestedTokenType["UrnDevrevParamsOauthTokenTypeSys"] = "urn:devrev:params:oauth:token-type:sys";
    AuthTokenRequestedTokenType["UrnIetfParamsOauthTokenTypeJwt"] = "urn:ietf:params:oauth:token-type:jwt";
})(AuthTokenRequestedTokenType = exports.AuthTokenRequestedTokenType || (exports.AuthTokenRequestedTokenType = {}));
/** The status of the token. */
var AuthTokenStatus;
(function (AuthTokenStatus) {
    AuthTokenStatus["Active"] = "active";
    AuthTokenStatus["Expired"] = "expired";
    AuthTokenStatus["Revoked"] = "revoked";
})(AuthTokenStatus = exports.AuthTokenStatus || (exports.AuthTokenStatus = {}));
/** The type of the subject token. */
var AuthTokenSubjectTokenType;
(function (AuthTokenSubjectTokenType) {
    AuthTokenSubjectTokenType["UrnDevrevParamsOauthTokenTypeJwtAuth0"] = "urn:devrev:params:oauth:token-type:jwt:auth0";
    AuthTokenSubjectTokenType["UrnDevrevParamsOauthTokenTypeJwtDev"] = "urn:devrev:params:oauth:token-type:jwt:dev";
    AuthTokenSubjectTokenType["UrnDevrevParamsOauthTokenTypeRat"] = "urn:devrev:params:oauth:token-type:rat";
    AuthTokenSubjectTokenType["UrnDevrevParamsOauthTokenTypeRev"] = "urn:devrev:params:oauth:token-type:rev";
    AuthTokenSubjectTokenType["UrnDevrevParamsOauthTokenTypeRevinfo"] = "urn:devrev:params:oauth:token-type:revinfo";
    AuthTokenSubjectTokenType["UrnDevrevParamsOauthTokenTypeSession"] = "urn:devrev:params:oauth:token-type:session";
    AuthTokenSubjectTokenType["UrnDevrevParamsOauthTokenTypeSysu"] = "urn:devrev:params:oauth:token-type:sysu";
    AuthTokenSubjectTokenType["UrnDevrevParamsOauthTokenTypeUserinfo"] = "urn:devrev:params:oauth:token-type:userinfo";
    AuthTokenSubjectTokenType["UrnIetfParamsOauthTokenTypeJwt"] = "urn:ietf:params:oauth:token-type:jwt";
})(AuthTokenSubjectTokenType = exports.AuthTokenSubjectTokenType || (exports.AuthTokenSubjectTokenType = {}));
/** The type of the issued token. Bearer is the only supported token type. */
var AuthTokenTokenType;
(function (AuthTokenTokenType) {
    AuthTokenTokenType["Bearer"] = "bearer";
})(AuthTokenTokenType = exports.AuthTokenTokenType || (exports.AuthTokenTokenType = {}));
var BooleanExpressionType;
(function (BooleanExpressionType) {
    BooleanExpressionType["And"] = "and";
    BooleanExpressionType["Not"] = "not";
    BooleanExpressionType["Or"] = "or";
    BooleanExpressionType["Primitive"] = "primitive";
})(BooleanExpressionType = exports.BooleanExpressionType || (exports.BooleanExpressionType = {}));
var ConversationsCreateRequestTypeValue;
(function (ConversationsCreateRequestTypeValue) {
    ConversationsCreateRequestTypeValue["Support"] = "support";
})(ConversationsCreateRequestTypeValue = exports.ConversationsCreateRequestTypeValue || (exports.ConversationsCreateRequestTypeValue = {}));
/** Type of date filter. */
var DateFilterType;
(function (DateFilterType) {
    DateFilterType["Preset"] = "preset";
    DateFilterType["Range"] = "range";
})(DateFilterType = exports.DateFilterType || (exports.DateFilterType = {}));
/** Type of date preset. */
var DateTimePresetType;
(function (DateTimePresetType) {
    DateTimePresetType["LastNDays"] = "last_n_days";
    DateTimePresetType["NextNDays"] = "next_n_days";
})(DateTimePresetType = exports.DateTimePresetType || (exports.DateTimePresetType = {}));
/**
 * Defines the type for the authentication connection. Different types of
 * authentication connections have different configuration parameters.
 */
var DevOrgAuthConnectionsCreateRequestType;
(function (DevOrgAuthConnectionsCreateRequestType) {
    DevOrgAuthConnectionsCreateRequestType["GoogleApps"] = "google_apps";
    DevOrgAuthConnectionsCreateRequestType["Oidc"] = "oidc";
    DevOrgAuthConnectionsCreateRequestType["Okta"] = "okta";
    DevOrgAuthConnectionsCreateRequestType["Samlp"] = "samlp";
    DevOrgAuthConnectionsCreateRequestType["Waad"] = "waad";
})(DevOrgAuthConnectionsCreateRequestType = exports.DevOrgAuthConnectionsCreateRequestType || (exports.DevOrgAuthConnectionsCreateRequestType = {}));
/**
 * Specifies the type for the authentication connection. Different types
 * of authentication connections have different configuration parameters
 * that can be updated.
 */
var DevOrgAuthConnectionsUpdateRequestType;
(function (DevOrgAuthConnectionsUpdateRequestType) {
    DevOrgAuthConnectionsUpdateRequestType["GoogleApps"] = "google_apps";
    DevOrgAuthConnectionsUpdateRequestType["None"] = "none";
    DevOrgAuthConnectionsUpdateRequestType["Oidc"] = "oidc";
    DevOrgAuthConnectionsUpdateRequestType["Okta"] = "okta";
    DevOrgAuthConnectionsUpdateRequestType["Samlp"] = "samlp";
    DevOrgAuthConnectionsUpdateRequestType["Waad"] = "waad";
})(DevOrgAuthConnectionsUpdateRequestType = exports.DevOrgAuthConnectionsUpdateRequestType || (exports.DevOrgAuthConnectionsUpdateRequestType = {}));
/** Job title of the Dev User. */
var DevUserJobTitle;
(function (DevUserJobTitle) {
    DevUserJobTitle["CustomerSuccessManager"] = "customer_success_manager";
    DevUserJobTitle["Cxo"] = "cxo";
    DevUserJobTitle["Designer"] = "designer";
    DevUserJobTitle["Developer"] = "developer";
    DevUserJobTitle["HeadOfSupport"] = "head_of_support";
    DevUserJobTitle["Operations"] = "operations";
    DevUserJobTitle["Others"] = "others";
    DevUserJobTitle["ProductManager"] = "product_manager";
    DevUserJobTitle["Qa"] = "qa";
    DevUserJobTitle["RevenueLeader"] = "revenue_leader";
    DevUserJobTitle["Support"] = "support";
    DevUserJobTitle["TechLead"] = "tech_lead";
})(DevUserJobTitle = exports.DevUserJobTitle || (exports.DevUserJobTitle = {}));
var ErrorBadRequestType;
(function (ErrorBadRequestType) {
    ErrorBadRequestType["ArtifactAlreadyAttachedToAParent"] = "artifact_already_attached_to_a_parent";
    ErrorBadRequestType["BadRequest"] = "bad_request";
    ErrorBadRequestType["InvalidApiVersion"] = "invalid_api_version";
    ErrorBadRequestType["InvalidEnumValue"] = "invalid_enum_value";
    ErrorBadRequestType["InvalidField"] = "invalid_field";
    ErrorBadRequestType["InvalidId"] = "invalid_id";
    ErrorBadRequestType["MergeWorksError"] = "merge_works_error";
    ErrorBadRequestType["MissingDependency"] = "missing_dependency";
    ErrorBadRequestType["MissingRequiredField"] = "missing_required_field";
    ErrorBadRequestType["ParseError"] = "parse_error";
    ErrorBadRequestType["StaleSchemaFragments"] = "stale_schema_fragments";
    ErrorBadRequestType["UnexpectedIdType"] = "unexpected_id_type";
    ErrorBadRequestType["UnexpectedJsonType"] = "unexpected_json_type";
    ErrorBadRequestType["ValueNotPermitted"] = "value_not_permitted";
})(ErrorBadRequestType = exports.ErrorBadRequestType || (exports.ErrorBadRequestType = {}));
var ErrorBadRequestUnexpectedJsonTypeType;
(function (ErrorBadRequestUnexpectedJsonTypeType) {
    ErrorBadRequestUnexpectedJsonTypeType["Array"] = "array";
    ErrorBadRequestUnexpectedJsonTypeType["Bool"] = "bool";
    ErrorBadRequestUnexpectedJsonTypeType["Null"] = "null";
    ErrorBadRequestUnexpectedJsonTypeType["Number"] = "number";
    ErrorBadRequestUnexpectedJsonTypeType["Object"] = "object";
    ErrorBadRequestUnexpectedJsonTypeType["String"] = "string";
})(ErrorBadRequestUnexpectedJsonTypeType = exports.ErrorBadRequestUnexpectedJsonTypeType || (exports.ErrorBadRequestUnexpectedJsonTypeType = {}));
var ErrorConflictType;
(function (ErrorConflictType) {
    ErrorConflictType["Conflict"] = "conflict";
})(ErrorConflictType = exports.ErrorConflictType || (exports.ErrorConflictType = {}));
var ErrorForbiddenType;
(function (ErrorForbiddenType) {
    ErrorForbiddenType["Forbidden"] = "forbidden";
})(ErrorForbiddenType = exports.ErrorForbiddenType || (exports.ErrorForbiddenType = {}));
var ErrorInternalServerErrorType;
(function (ErrorInternalServerErrorType) {
    ErrorInternalServerErrorType["InternalError"] = "internal_error";
})(ErrorInternalServerErrorType = exports.ErrorInternalServerErrorType || (exports.ErrorInternalServerErrorType = {}));
var ErrorNotFoundType;
(function (ErrorNotFoundType) {
    ErrorNotFoundType["NotFound"] = "not_found";
})(ErrorNotFoundType = exports.ErrorNotFoundType || (exports.ErrorNotFoundType = {}));
var ErrorServiceUnavailableType;
(function (ErrorServiceUnavailableType) {
    ErrorServiceUnavailableType["ServiceUnavailable"] = "service_unavailable";
})(ErrorServiceUnavailableType = exports.ErrorServiceUnavailableType || (exports.ErrorServiceUnavailableType = {}));
var ErrorTooManyRequestsType;
(function (ErrorTooManyRequestsType) {
    ErrorTooManyRequestsType["TooManyRequests"] = "too_many_requests";
})(ErrorTooManyRequestsType = exports.ErrorTooManyRequestsType || (exports.ErrorTooManyRequestsType = {}));
var ErrorUnauthorizedType;
(function (ErrorUnauthorizedType) {
    ErrorUnauthorizedType["Unauthenticated"] = "unauthenticated";
})(ErrorUnauthorizedType = exports.ErrorUnauthorizedType || (exports.ErrorUnauthorizedType = {}));
var EventFetchedResult;
(function (EventFetchedResult) {
    EventFetchedResult["Forbidden"] = "forbidden";
    EventFetchedResult["NotFound"] = "not_found";
    EventFetchedResult["Ok"] = "ok";
})(EventFetchedResult = exports.EventFetchedResult || (exports.EventFetchedResult = {}));
/** Creation source of the group. */
var GroupIngestionSource;
(function (GroupIngestionSource) {
    GroupIngestionSource["Airdrop"] = "airdrop";
    GroupIngestionSource["Scim"] = "scim";
})(GroupIngestionSource = exports.GroupIngestionSource || (exports.GroupIngestionSource = {}));
/** Type of the members in the group. */
var GroupMemberType;
(function (GroupMemberType) {
    GroupMemberType["DevUser"] = "dev_user";
    GroupMemberType["RevUser"] = "rev_user";
})(GroupMemberType = exports.GroupMemberType || (exports.GroupMemberType = {}));
/** Type of the group. */
var GroupType;
(function (GroupType) {
    GroupType["Dynamic"] = "dynamic";
    GroupType["Static"] = "static";
})(GroupType = exports.GroupType || (exports.GroupType = {}));
/** Priority of the work based upon impact and criticality. */
var IssuePriority;
(function (IssuePriority) {
    IssuePriority["P0"] = "p0";
    IssuePriority["P1"] = "p1";
    IssuePriority["P2"] = "p2";
    IssuePriority["P3"] = "p3";
})(IssuePriority = exports.IssuePriority || (exports.IssuePriority = {}));
var LinkEndpointType;
(function (LinkEndpointType) {
    LinkEndpointType["Capability"] = "capability";
    LinkEndpointType["Conversation"] = "conversation";
    LinkEndpointType["DevUser"] = "dev_user";
    LinkEndpointType["Enhancement"] = "enhancement";
    LinkEndpointType["Feature"] = "feature";
    LinkEndpointType["Issue"] = "issue";
    LinkEndpointType["Product"] = "product";
    LinkEndpointType["RevOrg"] = "rev_org";
    LinkEndpointType["RevUser"] = "rev_user";
    LinkEndpointType["Ticket"] = "ticket";
})(LinkEndpointType = exports.LinkEndpointType || (exports.LinkEndpointType = {}));
/** Type of link used to define the relationship. */
var LinkType;
(function (LinkType) {
    LinkType["CustomLink"] = "custom_link";
    LinkType["DevelopedWith"] = "developed_with";
    LinkType["Imports"] = "imports";
    LinkType["IsAnalyzedBy"] = "is_analyzed_by";
    LinkType["IsConvertedTo"] = "is_converted_to";
    LinkType["IsDependentOn"] = "is_dependent_on";
    LinkType["IsDuplicateOf"] = "is_duplicate_of";
    LinkType["IsFollowUpOf"] = "is_follow_up_of";
    LinkType["IsMergedInto"] = "is_merged_into";
    LinkType["IsParentOf"] = "is_parent_of";
    LinkType["IsPartOf"] = "is_part_of";
    LinkType["IsRelatedTo"] = "is_related_to";
    LinkType["Serves"] = "serves";
})(LinkType = exports.LinkType || (exports.LinkType = {}));
/**
 * The direction of link, which can either be outbound such that the
 * object is the source of the link, otherwise inbound where the object is
 * the target of the link.
 */
var LinksDirection;
(function (LinksDirection) {
    LinksDirection["IsSource"] = "is_source";
    LinksDirection["IsTarget"] = "is_target";
})(LinksDirection = exports.LinksDirection || (exports.LinksDirection = {}));
/**
 * The iteration mode to use. If "after", then entries after the provided
 * cursor will be returned, or if no cursor is provided, then from the
 * beginning. If "before", then entries before the provided cursor will be
 * returned, or if no cursor is provided, then from the end. Entries will
 * always be returned in the specified sort-by order.
 */
var ListMode;
(function (ListMode) {
    ListMode["After"] = "after";
    ListMode["Before"] = "before";
})(ListMode = exports.ListMode || (exports.ListMode = {}));
var MemberType;
(function (MemberType) {
    MemberType["DevUser"] = "dev_user";
    MemberType["RevUser"] = "rev_user";
    MemberType["SysUser"] = "sys_user";
})(MemberType = exports.MemberType || (exports.MemberType = {}));
/** The list of item types on which the metric might be applied. */
var MetricDefinitionAppliesTo;
(function (MetricDefinitionAppliesTo) {
    MetricDefinitionAppliesTo["Conversation"] = "conversation";
    MetricDefinitionAppliesTo["Issue"] = "issue";
    MetricDefinitionAppliesTo["Ticket"] = "ticket";
})(MetricDefinitionAppliesTo = exports.MetricDefinitionAppliesTo || (exports.MetricDefinitionAppliesTo = {}));
/**
 * The type of the metric. 'time' metrics track the time spent in some
 * state, supporting operations like 'start', 'stop', 'pause', 'resume'.
 * 'value' metrics track a value, supporting operations like 'set',
 * 'increment', 'decrement'.
 */
var MetricDefinitionMetricType;
(function (MetricDefinitionMetricType) {
    MetricDefinitionMetricType["Time"] = "time";
    MetricDefinitionMetricType["Value"] = "value";
})(MetricDefinitionMetricType = exports.MetricDefinitionMetricType || (exports.MetricDefinitionMetricType = {}));
/**
 * The status of the metric. 'active' metrics can be used to create new
 * SLAs, while 'inactive' metrics can not be used in new SLAs. Metrics can
 * be updated between 'active' and 'inactive' states.
 */
var MetricDefinitionStatus;
(function (MetricDefinitionStatus) {
    MetricDefinitionStatus["Active"] = "active";
    MetricDefinitionStatus["Inactive"] = "inactive";
})(MetricDefinitionStatus = exports.MetricDefinitionStatus || (exports.MetricDefinitionStatus = {}));
/** The environment of the Org. Defaults to 'production' if not specified. */
var OrgEnvironment;
(function (OrgEnvironment) {
    OrgEnvironment["Production"] = "production";
    OrgEnvironment["Staging"] = "staging";
    OrgEnvironment["Test"] = "test";
})(OrgEnvironment = exports.OrgEnvironment || (exports.OrgEnvironment = {}));
/**
 * Status determines how an item can be used. In 'draft' status an item
 * can be edited but can't be used. When 'published' the item can longer
 * be edited but can be used. 'Archived' is read-only.
 */
var OrgScheduleStatus;
(function (OrgScheduleStatus) {
    OrgScheduleStatus["Archived"] = "archived";
    OrgScheduleStatus["Draft"] = "draft";
    OrgScheduleStatus["Published"] = "published";
})(OrgScheduleStatus = exports.OrgScheduleStatus || (exports.OrgScheduleStatus = {}));
var OrgType;
(function (OrgType) {
    OrgType["RevOrg"] = "rev_org";
})(OrgType = exports.OrgType || (exports.OrgType = {}));
var PartType;
(function (PartType) {
    PartType["Capability"] = "capability";
    PartType["Enhancement"] = "enhancement";
    PartType["Feature"] = "feature";
    PartType["Product"] = "product";
})(PartType = exports.PartType || (exports.PartType = {}));
/** Status of the question answer. */
var QuestionAnswerStatus;
(function (QuestionAnswerStatus) {
    QuestionAnswerStatus["Archived"] = "archived";
    QuestionAnswerStatus["Discarded"] = "discarded";
    QuestionAnswerStatus["Draft"] = "draft";
    QuestionAnswerStatus["Published"] = "published";
    QuestionAnswerStatus["ReviewNeeded"] = "review_needed";
})(QuestionAnswerStatus = exports.QuestionAnswerStatus || (exports.QuestionAnswerStatus = {}));
var SchemaFieldDescriptorArrayTypeBaseType;
(function (SchemaFieldDescriptorArrayTypeBaseType) {
    SchemaFieldDescriptorArrayTypeBaseType["Bool"] = "bool";
    SchemaFieldDescriptorArrayTypeBaseType["Composite"] = "composite";
    SchemaFieldDescriptorArrayTypeBaseType["Date"] = "date";
    SchemaFieldDescriptorArrayTypeBaseType["Double"] = "double";
    SchemaFieldDescriptorArrayTypeBaseType["Enum"] = "enum";
    SchemaFieldDescriptorArrayTypeBaseType["Id"] = "id";
    SchemaFieldDescriptorArrayTypeBaseType["Int"] = "int";
    SchemaFieldDescriptorArrayTypeBaseType["JsonValue"] = "json_value";
    SchemaFieldDescriptorArrayTypeBaseType["RichText"] = "rich_text";
    SchemaFieldDescriptorArrayTypeBaseType["Struct"] = "struct";
    SchemaFieldDescriptorArrayTypeBaseType["Text"] = "text";
    SchemaFieldDescriptorArrayTypeBaseType["Timestamp"] = "timestamp";
    SchemaFieldDescriptorArrayTypeBaseType["Tokens"] = "tokens";
    SchemaFieldDescriptorArrayTypeBaseType["Uenum"] = "uenum";
})(SchemaFieldDescriptorArrayTypeBaseType = exports.SchemaFieldDescriptorArrayTypeBaseType || (exports.SchemaFieldDescriptorArrayTypeBaseType = {}));
var SchemaFieldDescriptorFieldType;
(function (SchemaFieldDescriptorFieldType) {
    SchemaFieldDescriptorFieldType["Array"] = "array";
    SchemaFieldDescriptorFieldType["Bool"] = "bool";
    SchemaFieldDescriptorFieldType["Composite"] = "composite";
    SchemaFieldDescriptorFieldType["Date"] = "date";
    SchemaFieldDescriptorFieldType["Double"] = "double";
    SchemaFieldDescriptorFieldType["Enum"] = "enum";
    SchemaFieldDescriptorFieldType["Id"] = "id";
    SchemaFieldDescriptorFieldType["Int"] = "int";
    SchemaFieldDescriptorFieldType["JsonValue"] = "json_value";
    SchemaFieldDescriptorFieldType["RichText"] = "rich_text";
    SchemaFieldDescriptorFieldType["Struct"] = "struct";
    SchemaFieldDescriptorFieldType["Text"] = "text";
    SchemaFieldDescriptorFieldType["Timestamp"] = "timestamp";
    SchemaFieldDescriptorFieldType["Tokens"] = "tokens";
    SchemaFieldDescriptorFieldType["Uenum"] = "uenum";
    SchemaFieldDescriptorFieldType["Unknown"] = "unknown";
})(SchemaFieldDescriptorFieldType = exports.SchemaFieldDescriptorFieldType || (exports.SchemaFieldDescriptorFieldType = {}));
/**
 * The object types on which the SLA applies. An external SLA can apply to
 * multiple object types, but an internal SLA can apply to only one object
 * type.
 */
var SlaAppliesTo;
(function (SlaAppliesTo) {
    SlaAppliesTo["Conversation"] = "conversation";
    SlaAppliesTo["Issue"] = "issue";
    SlaAppliesTo["Ticket"] = "ticket";
})(SlaAppliesTo = exports.SlaAppliesTo || (exports.SlaAppliesTo = {}));
/**
 * Specifies the frequency of when the contractually-meaningful evaluation
 * of the SLA happens. The organization for their own information might
 * evaluate the various metrics contained in the policy as often as we can
 * make it feasible for them, but only the evaluation at the end of the
 * evaluation period triggers a compensation and needs to be shared with
 * the customer.
 */
var SlaEvaluationPeriod;
(function (SlaEvaluationPeriod) {
    SlaEvaluationPeriod["Monthly"] = "monthly";
    SlaEvaluationPeriod["Quarterly"] = "quarterly";
    SlaEvaluationPeriod["Weekly"] = "weekly";
    SlaEvaluationPeriod["Yearly"] = "yearly";
})(SlaEvaluationPeriod = exports.SlaEvaluationPeriod || (exports.SlaEvaluationPeriod = {}));
/** The item type for which the SLA policy applies. */
var SlaSelectorAppliesTo;
(function (SlaSelectorAppliesTo) {
    SlaSelectorAppliesTo["Conversation"] = "conversation";
    SlaSelectorAppliesTo["Issue"] = "issue";
    SlaSelectorAppliesTo["Ticket"] = "ticket";
})(SlaSelectorAppliesTo = exports.SlaSelectorAppliesTo || (exports.SlaSelectorAppliesTo = {}));
/** The SLA policy applies to conversations with these priorities. */
var SlaSelectorPriority;
(function (SlaSelectorPriority) {
    SlaSelectorPriority["P0"] = "p0";
    SlaSelectorPriority["P1"] = "p1";
    SlaSelectorPriority["P2"] = "p2";
})(SlaSelectorPriority = exports.SlaSelectorPriority || (exports.SlaSelectorPriority = {}));
/** The SLA policy applies to tickets with these severities. */
var SlaSelectorSeverity;
(function (SlaSelectorSeverity) {
    SlaSelectorSeverity["Blocker"] = "blocker";
    SlaSelectorSeverity["High"] = "high";
    SlaSelectorSeverity["Low"] = "low";
    SlaSelectorSeverity["Medium"] = "medium";
})(SlaSelectorSeverity = exports.SlaSelectorSeverity || (exports.SlaSelectorSeverity = {}));
/**
 * Status determines how an item can be used. In 'draft' status an item
 * can be edited but can't be used. When 'published' the item can longer
 * be edited but can be used. 'Archived' is read-only.
 */
var SlaStatus;
(function (SlaStatus) {
    SlaStatus["Archived"] = "archived";
    SlaStatus["Draft"] = "draft";
    SlaStatus["Published"] = "published";
})(SlaStatus = exports.SlaStatus || (exports.SlaStatus = {}));
/**
 * The stage of the SLA. This is the metric stage which is closest to
 * breach.
 */
var SlaSummaryStage;
(function (SlaSummaryStage) {
    SlaSummaryStage["Breached"] = "breached";
    SlaSummaryStage["Completed"] = "completed";
    SlaSummaryStage["Paused"] = "paused";
    SlaSummaryStage["Running"] = "running";
    SlaSummaryStage["Warning"] = "warning";
})(SlaSummaryStage = exports.SlaSummaryStage || (exports.SlaSummaryStage = {}));
/** The type of the object on which the SLA is being tracked. */
var SlaTrackerAppliesToType;
(function (SlaTrackerAppliesToType) {
    SlaTrackerAppliesToType["Conversation"] = "conversation";
    SlaTrackerAppliesToType["Issue"] = "issue";
    SlaTrackerAppliesToType["Ticket"] = "ticket";
})(SlaTrackerAppliesToType = exports.SlaTrackerAppliesToType || (exports.SlaTrackerAppliesToType = {}));
/** Type of the SLA. */
var SlaType;
(function (SlaType) {
    SlaType["External"] = "external";
    SlaType["Internal"] = "internal";
})(SlaType = exports.SlaType || (exports.SlaType = {}));
var SlasFilterAppliesToOperatorType;
(function (SlasFilterAppliesToOperatorType) {
    SlasFilterAppliesToOperatorType["All"] = "all";
    SlasFilterAppliesToOperatorType["Any"] = "any";
})(SlasFilterAppliesToOperatorType = exports.SlasFilterAppliesToOperatorType || (exports.SlasFilterAppliesToOperatorType = {}));
/** Type of stage validation options when creating an object. */
var StageValidationOptionForCreate;
(function (StageValidationOptionForCreate) {
    StageValidationOptionForCreate["AllowNonStart"] = "allow_non_start";
})(StageValidationOptionForCreate = exports.StageValidationOptionForCreate || (exports.StageValidationOptionForCreate = {}));
/** Type of state validation options when updating the stage of an object. */
var StageValidationOptionForUpdate;
(function (StageValidationOptionForUpdate) {
    StageValidationOptionForUpdate["AllowInvalidTransition"] = "allow_invalid_transition";
})(StageValidationOptionForUpdate = exports.StageValidationOptionForUpdate || (exports.StageValidationOptionForUpdate = {}));
var SyncMetadataFilterSyncInFilterStatus;
(function (SyncMetadataFilterSyncInFilterStatus) {
    SyncMetadataFilterSyncInFilterStatus["Failed"] = "failed";
    SyncMetadataFilterSyncInFilterStatus["Modified"] = "modified";
    SyncMetadataFilterSyncInFilterStatus["Staged"] = "staged";
    SyncMetadataFilterSyncInFilterStatus["Succeeded"] = "succeeded";
})(SyncMetadataFilterSyncInFilterStatus = exports.SyncMetadataFilterSyncInFilterStatus || (exports.SyncMetadataFilterSyncInFilterStatus = {}));
var SyncMetadataFilterSyncOutFilterStatus;
(function (SyncMetadataFilterSyncOutFilterStatus) {
    SyncMetadataFilterSyncOutFilterStatus["Failed"] = "failed";
    SyncMetadataFilterSyncOutFilterStatus["Modified"] = "modified";
    SyncMetadataFilterSyncOutFilterStatus["Succeeded"] = "succeeded";
})(SyncMetadataFilterSyncOutFilterStatus = exports.SyncMetadataFilterSyncOutFilterStatus || (exports.SyncMetadataFilterSyncOutFilterStatus = {}));
/** The state of a sync progress. */
var SyncProgressState;
(function (SyncProgressState) {
    SyncProgressState["Completed"] = "completed";
    SyncProgressState["DeleteInProgress"] = "delete_in_progress";
    SyncProgressState["Deleted"] = "deleted";
    SyncProgressState["DeletionError"] = "deletion_error";
    SyncProgressState["ExtractAttachments"] = "extract_attachments";
    SyncProgressState["ExtractAttachmentsError"] = "extract_attachments_error";
    SyncProgressState["Extraction"] = "extraction";
    SyncProgressState["ExtractionError"] = "extraction_error";
    SyncProgressState["Loading"] = "loading";
    SyncProgressState["LoadingAttachments"] = "loading_attachments";
    SyncProgressState["LoadingAttachmentsError"] = "loading_attachments_error";
    SyncProgressState["LoadingError"] = "loading_error";
    SyncProgressState["RecipeDiscovery"] = "recipe_discovery";
    SyncProgressState["RecipeDiscoveryError"] = "recipe_discovery_error";
    SyncProgressState["RecipeDiscoveryWaitingForUserInput"] = "recipe_discovery_waiting_for_user_input";
    SyncProgressState["Starting"] = "starting";
    SyncProgressState["Transformation"] = "transformation";
    SyncProgressState["TransformationError"] = "transformation_error";
})(SyncProgressState = exports.SyncProgressState || (exports.SyncProgressState = {}));
/** The direction/mode of a sync run. */
var SyncRunMode;
(function (SyncRunMode) {
    SyncRunMode["Initial"] = "initial";
    SyncRunMode["SyncFromDevrev"] = "sync_from_devrev";
    SyncRunMode["SyncToDevrev"] = "sync_to_devrev";
})(SyncRunMode = exports.SyncRunMode || (exports.SyncRunMode = {}));
/** Channels of the ticket. */
var TicketChannels;
(function (TicketChannels) {
    TicketChannels["Email"] = "email";
    TicketChannels["Plug"] = "plug";
    TicketChannels["Slack"] = "slack";
    TicketChannels["Twilio"] = "twilio";
    TicketChannels["TwilioSms"] = "twilio_sms";
})(TicketChannels = exports.TicketChannels || (exports.TicketChannels = {}));
/** Severity of the ticket. */
var TicketSeverity;
(function (TicketSeverity) {
    TicketSeverity["Blocker"] = "blocker";
    TicketSeverity["High"] = "high";
    TicketSeverity["Low"] = "low";
    TicketSeverity["Medium"] = "medium";
})(TicketSeverity = exports.TicketSeverity || (exports.TicketSeverity = {}));
/** The type of the body to use for the comment. */
var TimelineCommentBodyType;
(function (TimelineCommentBodyType) {
    TimelineCommentBodyType["Data"] = "data";
    TimelineCommentBodyType["SnapKit"] = "snap_kit";
    TimelineCommentBodyType["SnapWidget"] = "snap_widget";
    TimelineCommentBodyType["Text"] = "text";
})(TimelineCommentBodyType = exports.TimelineCommentBodyType || (exports.TimelineCommentBodyType = {}));
var TimelineEntriesCreateRequestType;
(function (TimelineEntriesCreateRequestType) {
    TimelineEntriesCreateRequestType["TimelineComment"] = "timeline_comment";
})(TimelineEntriesCreateRequestType = exports.TimelineEntriesCreateRequestType || (exports.TimelineEntriesCreateRequestType = {}));
var TimelineEntriesUpdateRequestType;
(function (TimelineEntriesUpdateRequestType) {
    TimelineEntriesUpdateRequestType["TimelineComment"] = "timeline_comment";
})(TimelineEntriesUpdateRequestType = exports.TimelineEntriesUpdateRequestType || (exports.TimelineEntriesUpdateRequestType = {}));
/** The type of object that the Timeline entry belongs to. */
var TimelineEntryObjectType;
(function (TimelineEntryObjectType) {
    TimelineEntryObjectType["Capability"] = "capability";
    TimelineEntryObjectType["Conversation"] = "conversation";
    TimelineEntryObjectType["Enhancement"] = "enhancement";
    TimelineEntryObjectType["Feature"] = "feature";
    TimelineEntryObjectType["Issue"] = "issue";
    TimelineEntryObjectType["Product"] = "product";
    TimelineEntryObjectType["RevOrg"] = "rev_org";
    TimelineEntryObjectType["RevUser"] = "rev_user";
    TimelineEntryObjectType["Ticket"] = "ticket";
    TimelineEntryObjectType["TimelineComment"] = "timeline_comment";
})(TimelineEntryObjectType = exports.TimelineEntryObjectType || (exports.TimelineEntryObjectType = {}));
var TimelineEntryType;
(function (TimelineEntryType) {
    TimelineEntryType["TimelineComment"] = "timeline_comment";
})(TimelineEntryType = exports.TimelineEntryType || (exports.TimelineEntryType = {}));
/**
 * The visibility of the entry. If 'private', then the entry is only
 * visible to the creator, 'internal' is visible with the Dev
 * organization, 'external' is visible to the Dev organzation and Rev
 * users, and 'public' is visible to all. If not set, then the default
 * visibility is 'external'.
 */
var TimelineEntryVisibility;
(function (TimelineEntryVisibility) {
    TimelineEntryVisibility["External"] = "external";
    TimelineEntryVisibility["Internal"] = "internal";
    TimelineEntryVisibility["Private"] = "private";
    TimelineEntryVisibility["Public"] = "public";
})(TimelineEntryVisibility = exports.TimelineEntryVisibility || (exports.TimelineEntryVisibility = {}));
/** State of the user. */
var UserState;
(function (UserState) {
    UserState["Active"] = "active";
    UserState["Deactivated"] = "deactivated";
    UserState["Deleted"] = "deleted";
    UserState["Locked"] = "locked";
    UserState["Shadow"] = "shadow";
    UserState["Unassigned"] = "unassigned";
})(UserState = exports.UserState || (exports.UserState = {}));
var UserType;
(function (UserType) {
    UserType["DevUser"] = "dev_user";
    UserType["RevUser"] = "rev_user";
    UserType["SysUser"] = "sys_user";
})(UserType = exports.UserType || (exports.UserType = {}));
/** Defines the state of the group item. */
var VistaGroupItemState;
(function (VistaGroupItemState) {
    VistaGroupItemState["Active"] = "active";
    VistaGroupItemState["Completed"] = "completed";
    VistaGroupItemState["Planned"] = "planned";
})(VistaGroupItemState = exports.VistaGroupItemState || (exports.VistaGroupItemState = {}));
/** Type of the group object. */
var VistaGroupItemType;
(function (VistaGroupItemType) {
    VistaGroupItemType["Curated"] = "curated";
    VistaGroupItemType["Dynamic"] = "dynamic";
})(VistaGroupItemType = exports.VistaGroupItemType || (exports.VistaGroupItemType = {}));
/** The event types that the webhook will receive. */
var WebhookEventType;
(function (WebhookEventType) {
    WebhookEventType["AccountCreated"] = "account_created";
    WebhookEventType["AccountDeleted"] = "account_deleted";
    WebhookEventType["AccountUpdated"] = "account_updated";
    WebhookEventType["ConversationCreated"] = "conversation_created";
    WebhookEventType["ConversationDeleted"] = "conversation_deleted";
    WebhookEventType["ConversationUpdated"] = "conversation_updated";
    WebhookEventType["DevUserCreated"] = "dev_user_created";
    WebhookEventType["DevUserDeleted"] = "dev_user_deleted";
    WebhookEventType["DevUserUpdated"] = "dev_user_updated";
    WebhookEventType["GroupCreated"] = "group_created";
    WebhookEventType["GroupDeleted"] = "group_deleted";
    WebhookEventType["GroupMemberAdded"] = "group_member_added";
    WebhookEventType["GroupMemberRemoved"] = "group_member_removed";
    WebhookEventType["GroupUpdated"] = "group_updated";
    WebhookEventType["IncidentCreated"] = "incident_created";
    WebhookEventType["IncidentDeleted"] = "incident_deleted";
    WebhookEventType["IncidentUpdated"] = "incident_updated";
    WebhookEventType["LinkCreated"] = "link_created";
    WebhookEventType["LinkDeleted"] = "link_deleted";
    WebhookEventType["LinkUpdated"] = "link_updated";
    WebhookEventType["PartCreated"] = "part_created";
    WebhookEventType["PartDeleted"] = "part_deleted";
    WebhookEventType["PartUpdated"] = "part_updated";
    WebhookEventType["QuestionAnswerCreated"] = "question_answer_created";
    WebhookEventType["QuestionAnswerDeleted"] = "question_answer_deleted";
    WebhookEventType["QuestionAnswerUpdated"] = "question_answer_updated";
    WebhookEventType["RevOrgCreated"] = "rev_org_created";
    WebhookEventType["RevOrgDeleted"] = "rev_org_deleted";
    WebhookEventType["RevOrgUpdated"] = "rev_org_updated";
    WebhookEventType["RevUserCreated"] = "rev_user_created";
    WebhookEventType["RevUserDeleted"] = "rev_user_deleted";
    WebhookEventType["RevUserUpdated"] = "rev_user_updated";
    WebhookEventType["SlaTrackerCreated"] = "sla_tracker_created";
    WebhookEventType["SlaTrackerDeleted"] = "sla_tracker_deleted";
    WebhookEventType["SlaTrackerUpdated"] = "sla_tracker_updated";
    WebhookEventType["SurveyCreated"] = "survey_created";
    WebhookEventType["SurveyDeleted"] = "survey_deleted";
    WebhookEventType["SurveyResponseCreated"] = "survey_response_created";
    WebhookEventType["SurveyResponseDeleted"] = "survey_response_deleted";
    WebhookEventType["SurveyResponseUpdated"] = "survey_response_updated";
    WebhookEventType["SurveyUpdated"] = "survey_updated";
    WebhookEventType["SyncHistoryCreated"] = "sync_history_created";
    WebhookEventType["SyncHistoryDeleted"] = "sync_history_deleted";
    WebhookEventType["SyncHistoryUpdated"] = "sync_history_updated";
    WebhookEventType["SyncUnitUpdated"] = "sync_unit_updated";
    WebhookEventType["TagCreated"] = "tag_created";
    WebhookEventType["TagDeleted"] = "tag_deleted";
    WebhookEventType["TagUpdated"] = "tag_updated";
    WebhookEventType["TimelineEntryCreated"] = "timeline_entry_created";
    WebhookEventType["TimelineEntryDeleted"] = "timeline_entry_deleted";
    WebhookEventType["TimelineEntryUpdated"] = "timeline_entry_updated";
    WebhookEventType["Verify"] = "verify";
    WebhookEventType["WebhookCreated"] = "webhook_created";
    WebhookEventType["WebhookDeleted"] = "webhook_deleted";
    WebhookEventType["WebhookUpdated"] = "webhook_updated";
    WebhookEventType["WorkCreated"] = "work_created";
    WebhookEventType["WorkDeleted"] = "work_deleted";
    WebhookEventType["WorkFetched"] = "work_fetched";
    WebhookEventType["WorkUpdated"] = "work_updated";
})(WebhookEventType = exports.WebhookEventType || (exports.WebhookEventType = {}));
/** The status of the webhook. */
var WebhookStatus;
(function (WebhookStatus) {
    WebhookStatus["Active"] = "active";
    WebhookStatus["Inactive"] = "inactive";
    WebhookStatus["Unverified"] = "unverified";
})(WebhookStatus = exports.WebhookStatus || (exports.WebhookStatus = {}));
/** The action to update the webhook's status. */
var WebhooksUpdateAction;
(function (WebhooksUpdateAction) {
    WebhooksUpdateAction["Activate"] = "activate";
    WebhooksUpdateAction["Deactivate"] = "deactivate";
})(WebhooksUpdateAction = exports.WebhooksUpdateAction || (exports.WebhooksUpdateAction = {}));
var WorkType;
(function (WorkType) {
    WorkType["Issue"] = "issue";
    WorkType["Ticket"] = "ticket";
})(WorkType = exports.WorkType || (exports.WorkType = {}));
const axios_1 = __importDefault(require("axios"));
var ContentType;
(function (ContentType) {
    ContentType["Json"] = "application/json";
    ContentType["FormData"] = "multipart/form-data";
    ContentType["UrlEncoded"] = "application/x-www-form-urlencoded";
    ContentType["Text"] = "text/plain";
})(ContentType = exports.ContentType || (exports.ContentType = {}));
class HttpClient {
    constructor(_a = {}) {
        var { securityWorker, secure, format } = _a, axiosConfig = __rest(_a, ["securityWorker", "secure", "format"]);
        this.securityData = null;
        this.setSecurityData = (data) => {
            this.securityData = data;
        };
        this.request = (_b) => __awaiter(this, void 0, void 0, function* () {
            var { secure, path, type, query, format, body } = _b, params = __rest(_b, ["secure", "path", "type", "query", "format", "body"]);
            const secureParams = ((typeof secure === 'boolean' ? secure : this.secure) &&
                this.securityWorker &&
                (yield this.securityWorker(this.securityData))) ||
                {};
            const requestParams = this.mergeRequestParams(params, secureParams);
            const responseFormat = format || this.format || undefined;
            if (type === ContentType.FormData &&
                body &&
                body !== null &&
                typeof body === 'object') {
                body = this.createFormData(body);
            }
            if (type === ContentType.Text &&
                body &&
                body !== null &&
                typeof body !== 'string') {
                body = JSON.stringify(body);
            }
            return this.instance.request(Object.assign(Object.assign({}, requestParams), { headers: Object.assign(Object.assign({}, (requestParams.headers || {})), (type && type !== ContentType.FormData
                    ? { 'Content-Type': type }
                    : {})), params: query, responseType: responseFormat, data: body, url: path }));
        });
        this.instance = axios_1.default.create(Object.assign(Object.assign({}, axiosConfig), { baseURL: axiosConfig.baseURL || '{protocol}://{hostname}' }));
        this.secure = secure;
        this.format = format;
        this.securityWorker = securityWorker;
    }
    mergeRequestParams(params1, params2) {
        const method = params1.method || (params2 && params2.method);
        return Object.assign(Object.assign(Object.assign(Object.assign({}, this.instance.defaults), params1), (params2 || {})), { headers: Object.assign(Object.assign(Object.assign({}, ((method &&
                this.instance.defaults.headers[method.toLowerCase()]) ||
                {})), (params1.headers || {})), ((params2 && params2.headers) || {})) });
    }
    stringifyFormItem(formItem) {
        if (typeof formItem === 'object' && formItem !== null) {
            return JSON.stringify(formItem);
        }
        else {
            return `${formItem}`;
        }
    }
    createFormData(input) {
        return Object.keys(input || {}).reduce((formData, key) => {
            const property = input[key];
            const propertyContent = property instanceof Array ? property : [property];
            for (const formItem of propertyContent) {
                const isFileType = formItem instanceof Blob || formItem instanceof File;
                formData.append(key, isFileType ? formItem : this.stringifyFormItem(formItem));
            }
            return formData;
        }, new FormData());
    }
}
exports.HttpClient = HttpClient;
/**
 * @title DevRev REST API
 * @version 2022-10-20
 * @baseUrl {protocol}://{hostname}
 *
 * DevRev's REST API.
 */
class Api extends HttpClient {
    constructor() {
        super(...arguments);
        /**
         * @description Creates an account, which is a record representing a customer or an organization.
         *
         * @tags accounts
         * @name AccountsCreate
         * @summary Create Account
         * @request POST:/accounts.create
         * @secure
         */
        this.accountsCreate = (data, params = {}) => this.request(Object.assign({ path: `/accounts.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes an account.
         *
         * @tags accounts
         * @name AccountsDelete
         * @summary Delete Account
         * @request POST:/accounts.delete
         * @secure
         */
        this.accountsDelete = (data, params = {}) => this.request(Object.assign({ path: `/accounts.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Exports a collection of accounts.
         *
         * @tags accounts
         * @name AccountsExport
         * @summary Export Accounts
         * @request GET:/accounts.export
         * @secure
         */
        this.accountsExport = (query, params = {}) => this.request(Object.assign({ path: `/accounts.export`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Exports a collection of accounts.
         *
         * @tags accounts
         * @name AccountsExportPost
         * @summary Export Accounts (POST)
         * @request POST:/accounts.export
         * @secure
         */
        this.accountsExportPost = (data, params = {}) => this.request(Object.assign({ path: `/accounts.export`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Retrieves an account's information.
         *
         * @tags accounts
         * @name AccountsGet
         * @summary Get Account
         * @request GET:/accounts.get
         * @secure
         */
        this.accountsGet = (query, params = {}) => this.request(Object.assign({ path: `/accounts.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Retrieves an account's information.
         *
         * @tags accounts
         * @name AccountsGetPost
         * @summary Get Account (POST)
         * @request POST:/accounts.get
         * @secure
         */
        this.accountsGetPost = (data, params = {}) => this.request(Object.assign({ path: `/accounts.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets a list of accounts.
         *
         * @tags accounts
         * @name AccountsList
         * @summary List Accounts
         * @request GET:/accounts.list
         * @secure
         */
        this.accountsList = (query, params = {}) => this.request(Object.assign({ path: `/accounts.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets a list of accounts.
         *
         * @tags accounts
         * @name AccountsListPost
         * @summary List Accounts (POST)
         * @request POST:/accounts.list
         * @secure
         */
        this.accountsListPost = (data, params = {}) => this.request(Object.assign({ path: `/accounts.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates an account's information.
         *
         * @tags accounts
         * @name AccountsUpdate
         * @summary Update Account
         * @request POST:/accounts.update
         * @secure
         */
        this.accountsUpdate = (data, params = {}) => this.request(Object.assign({ path: `/accounts.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Article is an object which can contain a URL or artifacts in the resource. It also contains the data regarding the owner, author, status and published date of the object. This call creates an article.
         *
         * @tags articles
         * @name CreateArticle
         * @summary Create Article
         * @request POST:/articles.create
         * @secure
         */
        this.createArticle = (data, params = {}) => this.request(Object.assign({ path: `/articles.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes an article.
         *
         * @tags articles
         * @name DeleteArticle
         * @summary Delete Article
         * @request POST:/articles.delete
         * @secure
         */
        this.deleteArticle = (data, params = {}) => this.request(Object.assign({ path: `/articles.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets an article.
         *
         * @tags articles
         * @name GetArticle
         * @summary Get Article
         * @request GET:/articles.get
         * @secure
         */
        this.getArticle = (query, params = {}) => this.request(Object.assign({ path: `/articles.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets an article.
         *
         * @tags articles
         * @name GetArticlePost
         * @summary Get Article (POST)
         * @request POST:/articles.get
         * @secure
         */
        this.getArticlePost = (data, params = {}) => this.request(Object.assign({ path: `/articles.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists a collection of articles.
         *
         * @tags articles
         * @name ListArticles
         * @summary List Articles
         * @request GET:/articles.list
         * @secure
         */
        this.listArticles = (query, params = {}) => this.request(Object.assign({ path: `/articles.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists a collection of articles.
         *
         * @tags articles
         * @name ListArticlesPost
         * @summary List Articles (POST)
         * @request POST:/articles.list
         * @secure
         */
        this.listArticlesPost = (data, params = {}) => this.request(Object.assign({ path: `/articles.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates an article.
         *
         * @tags articles
         * @name UpdateArticle
         * @summary Update Article
         * @request POST:/articles.update
         * @secure
         */
        this.updateArticle = (data, params = {}) => this.request(Object.assign({ path: `/articles.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the requested artifact's information.
         *
         * @tags artifacts
         * @name ArtifactsGet
         * @summary Get Artifact
         * @request GET:/artifacts.get
         * @secure
         */
        this.artifactsGet = (query, params = {}) => this.request(Object.assign({ path: `/artifacts.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the requested artifact's information.
         *
         * @tags artifacts
         * @name ArtifactsGetPost
         * @summary Get Artifact (POST)
         * @request POST:/artifacts.get
         * @secure
         */
        this.artifactsGetPost = (data, params = {}) => this.request(Object.assign({ path: `/artifacts.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description List the artifacts attached to an object.
         *
         * @tags artifacts
         * @name ArtifactsList
         * @summary List Artifacts
         * @request GET:/artifacts.list
         * @secure
         */
        this.artifactsList = (query, params = {}) => this.request(Object.assign({ path: `/artifacts.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description List the artifacts attached to an object.
         *
         * @tags artifacts
         * @name ArtifactsListPost
         * @summary List Artifacts (POST)
         * @request POST:/artifacts.list
         * @secure
         */
        this.artifactsListPost = (data, params = {}) => this.request(Object.assign({ path: `/artifacts.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the download URL for the artifact.
         *
         * @tags artifacts
         * @name ArtifactsLocate
         * @summary Locate Artifacts
         * @request GET:/artifacts.locate
         * @secure
         */
        this.artifactsLocate = (query, params = {}) => this.request(Object.assign({ path: `/artifacts.locate`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the download URL for the artifact.
         *
         * @tags artifacts
         * @name ArtifactsLocatePost
         * @summary Locate Artifacts (POST)
         * @request POST:/artifacts.locate
         * @secure
         */
        this.artifactsLocatePost = (data, params = {}) => this.request(Object.assign({ path: `/artifacts.locate`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates an artifact and generates an upload URL for its data.
         *
         * @tags artifacts
         * @name ArtifactsPrepare
         * @summary Prepare Artifacts
         * @request POST:/artifacts.prepare
         * @secure
         */
        this.artifactsPrepare = (data, params = {}) => this.request(Object.assign({ path: `/artifacts.prepare`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the specified object.
         *
         * @tags atoms
         * @name AtomsGet
         * @summary Get Atom
         * @request GET:/atoms.get
         * @secure
         */
        this.atomsGet = (query, params = {}) => this.request(Object.assign({ path: `/atoms.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the specified object.
         *
         * @tags atoms
         * @name AtomsGetPost
         * @summary Get Atom (POST)
         * @request POST:/atoms.get
         * @secure
         */
        this.atomsGetPost = (data, params = {}) => this.request(Object.assign({ path: `/atoms.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates a JWT corresponding to the requested token type for the authenticated user.
         *
         * @tags auth-tokens
         * @name AuthTokensCreate
         * @summary Create Auth Token
         * @request POST:/auth-tokens.create
         * @secure
         */
        this.authTokensCreate = (data, params = {}) => this.request(Object.assign({ path: `/auth-tokens.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Revokes the token that matches the given token ID issued under the given Dev organization.
         *
         * @tags auth-tokens
         * @name AuthTokensDelete
         * @summary Delete Auth Token
         * @request POST:/auth-tokens.delete
         * @secure
         */
        this.authTokensDelete = (data, params = {}) => this.request(Object.assign({ path: `/auth-tokens.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json }, params));
        /**
         * @description Gets the token metadata corresponding to the given token ID under the given Dev organization.
         *
         * @tags auth-tokens
         * @name AuthTokensGet
         * @summary Get Auth Token
         * @request GET:/auth-tokens.get
         * @secure
         */
        this.authTokensGet = (query, params = {}) => this.request(Object.assign({ path: `/auth-tokens.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the token metadata corresponding to the given token ID under the given Dev organization.
         *
         * @tags auth-tokens
         * @name AuthTokensGetPost
         * @summary Get Auth Token (POST)
         * @request POST:/auth-tokens.get
         * @secure
         */
        this.authTokensGetPost = (data, params = {}) => this.request(Object.assign({ path: `/auth-tokens.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the token metadata for all the tokens corresponding to the given token type issued for a given subject.
         *
         * @tags auth-tokens
         * @name AuthTokensList
         * @summary List Auth Tokens
         * @request GET:/auth-tokens.list
         * @secure
         */
        this.authTokensList = (query, params = {}) => this.request(Object.assign({ path: `/auth-tokens.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the token metadata for all the tokens corresponding to the given token type issued for a given subject.
         *
         * @tags auth-tokens
         * @name AuthTokensListPost
         * @summary List Auth Tokens (POST)
         * @request POST:/auth-tokens.list
         * @secure
         */
        this.authTokensListPost = (data, params = {}) => this.request(Object.assign({ path: `/auth-tokens.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Revokes all the tokens that matches the given token type created by the authenticated user.
         *
         * @tags auth-tokens
         * @name AuthTokensSelfDelete
         * @summary Delete Auth Tokens Self
         * @request POST:/auth-tokens.self.delete
         * @secure
         */
        this.authTokensSelfDelete = (data, params = {}) => this.request(Object.assign({ path: `/auth-tokens.self.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json }, params));
        /**
         * @description Updates token metadata of a token issued under a given Dev organization.
         *
         * @tags auth-tokens
         * @name AuthTokensUpdate
         * @summary Update Auth Token
         * @request POST:/auth-tokens.update
         * @secure
         */
        this.authTokensUpdate = (data, params = {}) => this.request(Object.assign({ path: `/auth-tokens.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates a conversation.
         *
         * @tags conversations
         * @name ConversationsCreate
         * @summary Create Conversation
         * @request POST:/conversations.create
         * @secure
         */
        this.conversationsCreate = (data, params = {}) => this.request(Object.assign({ path: `/conversations.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes the requested conversation.
         *
         * @tags conversations
         * @name ConversationsDelete
         * @summary Delete Conversation
         * @request POST:/conversations.delete
         * @secure
         */
        this.conversationsDelete = (data, params = {}) => this.request(Object.assign({ path: `/conversations.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the requested conversation's information.
         *
         * @tags conversations
         * @name ConversationsGet
         * @summary Get Conversation
         * @request GET:/conversations.get
         * @secure
         */
        this.conversationsGet = (query, params = {}) => this.request(Object.assign({ path: `/conversations.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the requested conversation's information.
         *
         * @tags conversations
         * @name ConversationsGetPost
         * @summary Get Conversation (POST)
         * @request POST:/conversations.get
         * @secure
         */
        this.conversationsGetPost = (data, params = {}) => this.request(Object.assign({ path: `/conversations.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists the available conversations.
         *
         * @tags conversations
         * @name ConversationsList
         * @summary List Conversations
         * @request GET:/conversations.list
         * @secure
         */
        this.conversationsList = (query, params = {}) => this.request(Object.assign({ path: `/conversations.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists the available conversations.
         *
         * @tags conversations
         * @name ConversationsListPost
         * @summary List Conversations (POST)
         * @request POST:/conversations.list
         * @secure
         */
        this.conversationsListPost = (data, params = {}) => this.request(Object.assign({ path: `/conversations.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates the requested conversation.
         *
         * @tags conversations
         * @name ConversationsUpdate
         * @summary Update Conversation
         * @request POST:/conversations.update
         * @secure
         */
        this.conversationsUpdate = (data, params = {}) => this.request(Object.assign({ path: `/conversations.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates a new enterprise authentication connection for a Dev organization. This authentication connection will not be enabled by default for the organization and the user will need to explicitly enable this. Keep in mind that at a time, only one authentication connection can be enabled for a Dev organization. At present, only 5 enterprise connections can be created by an organization.
         *
         * @tags auth-connections, dev-orgs
         * @name DevOrgAuthConnectionsCreate
         * @summary Create Dev Orgs Auth Connection
         * @request POST:/dev-orgs.auth-connections.create
         * @secure
         */
        this.devOrgAuthConnectionsCreate = (data, params = {}) => this.request(Object.assign({ path: `/dev-orgs.auth-connections.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes an authentication connection. Only enterprise connections which are explicitly set up for a Dev organization can be deleted. Default connections can not be deleted using this method.
         *
         * @tags auth-connections, dev-orgs
         * @name DevOrgAuthConnectionsDelete
         * @summary Delete Dev Orgs Auth Connection
         * @request POST:/dev-orgs.auth-connections.delete
         * @secure
         */
        this.devOrgAuthConnectionsDelete = (data, params = {}) => this.request(Object.assign({ path: `/dev-orgs.auth-connections.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json }, params));
        /**
         * @description Retrieves the details for an authentication connection.
         *
         * @tags auth-connections, dev-orgs
         * @name DevOrgAuthConnectionsGet
         * @summary Get Dev Orgs Auth Connection
         * @request GET:/dev-orgs.auth-connections.get
         * @secure
         */
        this.devOrgAuthConnectionsGet = (query, params = {}) => this.request(Object.assign({ path: `/dev-orgs.auth-connections.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Retrieves the details for an authentication connection.
         *
         * @tags auth-connections, dev-orgs
         * @name DevOrgAuthConnectionsGetPost
         * @summary Get Dev Orgs Auth Connection (POST)
         * @request POST:/dev-orgs.auth-connections.get
         * @secure
         */
        this.devOrgAuthConnectionsGetPost = (data, params = {}) => this.request(Object.assign({ path: `/dev-orgs.auth-connections.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists all the authentication connections available for a Dev organization. This list will include both social and enterprise connections which are either available by default or are explicitly created by the user.
         *
         * @tags auth-connections, dev-orgs
         * @name DevOrgAuthConnectionsList
         * @summary List Dev Orgs Auth Connections
         * @request GET:/dev-orgs.auth-connections.list
         * @secure
         */
        this.devOrgAuthConnectionsList = (params = {}) => this.request(Object.assign({ path: `/dev-orgs.auth-connections.list`, method: 'GET', secure: true, format: 'json' }, params));
        /**
         * @description Lists all the authentication connections available for a Dev organization. This list will include both social and enterprise connections which are either available by default or are explicitly created by the user.
         *
         * @tags auth-connections, dev-orgs
         * @name DevOrgAuthConnectionsListPost
         * @summary List Dev Orgs Auth Connections (POST)
         * @request POST:/dev-orgs.auth-connections.list
         * @secure
         */
        this.devOrgAuthConnectionsListPost = (data, params = {}) => this.request(Object.assign({ path: `/dev-orgs.auth-connections.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Enable or disable an authentication connection for a Dev organization. Currently, only 1 authentication connection can be enabled at a time. When a new authentication connection is enabled, the connection which is currently enabled for the Dev organization is automatically disabled.
         *
         * @tags auth-connections, dev-orgs
         * @name DevOrgAuthConnectionsToggle
         * @summary Toggle Dev Orgs Auth Connections
         * @request POST:/dev-orgs.auth-connections.toggle
         * @secure
         */
        this.devOrgAuthConnectionsToggle = (data, params = {}) => this.request(Object.assign({ path: `/dev-orgs.auth-connections.toggle`, method: 'POST', body: data, secure: true, type: ContentType.Json }, params));
        /**
         * @description Updates an authentication connection.
         *
         * @tags auth-connections, dev-orgs
         * @name DevOrgAuthConnectionsUpdate
         * @summary Update Dev Orgs Auth Connection
         * @request POST:/dev-orgs.auth-connections.update
         * @secure
         */
        this.devOrgAuthConnectionsUpdate = (data, params = {}) => this.request(Object.assign({ path: `/dev-orgs.auth-connections.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the requested user's information.
         *
         * @tags dev-users
         * @name DevUsersGet
         * @summary Get Dev User
         * @request GET:/dev-users.get
         * @secure
         */
        this.devUsersGet = (query, params = {}) => this.request(Object.assign({ path: `/dev-users.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the requested user's information.
         *
         * @tags dev-users
         * @name DevUsersGetPost
         * @summary Get Dev User (POST)
         * @request POST:/dev-users.get
         * @secure
         */
        this.devUsersGetPost = (data, params = {}) => this.request(Object.assign({ path: `/dev-users.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists users within your organization.
         *
         * @tags dev-users
         * @name DevUsersList
         * @summary List Dev Users
         * @request GET:/dev-users.list
         * @secure
         */
        this.devUsersList = (query, params = {}) => this.request(Object.assign({ path: `/dev-users.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists users within your organization.
         *
         * @tags dev-users
         * @name DevUsersListPost
         * @summary List Dev Users (POST)
         * @request POST:/dev-users.list
         * @secure
         */
        this.devUsersListPost = (data, params = {}) => this.request(Object.assign({ path: `/dev-users.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the authenticated user's information.
         *
         * @tags dev-users
         * @name DevUsersSelf
         * @summary Self Dev User
         * @request GET:/dev-users.self
         * @secure
         */
        this.devUsersSelf = (params = {}) => this.request(Object.assign({ path: `/dev-users.self`, method: 'GET', secure: true, format: 'json' }, params));
        /**
         * @description Gets the authenticated user's information.
         *
         * @tags dev-users
         * @name DevUsersSelfPost
         * @summary Self Dev User (POST)
         * @request POST:/dev-users.self
         * @secure
         */
        this.devUsersSelfPost = (data, params = {}) => this.request(Object.assign({ path: `/dev-users.self`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates the authenticated user.
         *
         * @tags dev-users
         * @name DevUsersSelfUpdate
         * @summary Update Dev Users Self
         * @request POST:/dev-users.self.update
         * @secure
         */
        this.devUsersSelfUpdate = (data, params = {}) => this.request(Object.assign({ path: `/dev-users.self.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates the user corresponding to the input Id.
         *
         * @tags dev-users
         * @name DevUsersUpdate
         * @summary Update Dev User
         * @request POST:/dev-users.update
         * @secure
         */
        this.devUsersUpdate = (data, params = {}) => this.request(Object.assign({ path: `/dev-users.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates a new group. A group is a collection of users.
         *
         * @tags groups
         * @name GroupsCreate
         * @summary Create Group
         * @request POST:/groups.create
         * @secure
         */
        this.groupsCreate = (data, params = {}) => this.request(Object.assign({ path: `/groups.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the requested group.
         *
         * @tags groups
         * @name GroupsGet
         * @summary Get Group
         * @request GET:/groups.get
         * @secure
         */
        this.groupsGet = (query, params = {}) => this.request(Object.assign({ path: `/groups.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the requested group.
         *
         * @tags groups
         * @name GroupsGetPost
         * @summary Get Group (POST)
         * @request POST:/groups.get
         * @secure
         */
        this.groupsGetPost = (data, params = {}) => this.request(Object.assign({ path: `/groups.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists the available groups.
         *
         * @tags groups
         * @name GroupsList
         * @summary List Groups
         * @request GET:/groups.list
         * @secure
         */
        this.groupsList = (query, params = {}) => this.request(Object.assign({ path: `/groups.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists the available groups.
         *
         * @tags groups
         * @name GroupsListPost
         * @summary List Groups (POST)
         * @request POST:/groups.list
         * @secure
         */
        this.groupsListPost = (data, params = {}) => this.request(Object.assign({ path: `/groups.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Adds a member to a group.
         *
         * @tags groups
         * @name GroupMembersAdd
         * @summary Add Groups Members
         * @request POST:/groups.members.add
         * @secure
         */
        this.groupMembersAdd = (data, params = {}) => this.request(Object.assign({ path: `/groups.members.add`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists the members in a group.
         *
         * @tags groups
         * @name GroupMembersList
         * @summary List Groups Members
         * @request GET:/groups.members.list
         * @secure
         */
        this.groupMembersList = (query, params = {}) => this.request(Object.assign({ path: `/groups.members.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists the members in a group.
         *
         * @tags groups
         * @name GroupMembersListPost
         * @summary List Groups Members (POST)
         * @request POST:/groups.members.list
         * @secure
         */
        this.groupMembersListPost = (data, params = {}) => this.request(Object.assign({ path: `/groups.members.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Removes a member from a group.
         *
         * @tags groups
         * @name GroupMembersRemove
         * @summary Remove Groups Members
         * @request POST:/groups.members.remove
         * @secure
         */
        this.groupMembersRemove = (data, params = {}) => this.request(Object.assign({ path: `/groups.members.remove`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates the requested group.
         *
         * @tags groups
         * @name GroupsUpdate
         * @summary Update Group
         * @request POST:/groups.update
         * @secure
         */
        this.groupsUpdate = (data, params = {}) => this.request(Object.assign({ path: `/groups.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description OAuth2 authorization callback.
         *
         * @tags keyrings
         * @name KeyringsCreateCallback
         * @summary Authorize Keyrings
         * @request GET:/keyrings.authorize
         */
        this.keyringsCreateCallback = (query, params = {}) => this.request(Object.assign({ path: `/keyrings.authorize`, method: 'GET', query: query }, params));
        /**
         * @description OAuth2 authorization callback.
         *
         * @tags keyrings
         * @name KeyringsCreateCallbackPost
         * @summary Authorize Keyrings (POST)
         * @request POST:/keyrings.authorize
         * @secure
         */
        this.keyringsCreateCallbackPost = (data, params = {}) => this.request(Object.assign({ path: `/keyrings.authorize`, method: 'POST', body: data, secure: true, type: ContentType.Json }, params));
        /**
         * @description Creates a link between two objects to indicate a relationship.
         *
         * @tags links
         * @name LinksCreate
         * @summary Create Link
         * @request POST:/links.create
         * @secure
         */
        this.linksCreate = (data, params = {}) => this.request(Object.assign({ path: `/links.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes a link.
         *
         * @tags links
         * @name LinksDelete
         * @summary Delete Link
         * @request POST:/links.delete
         * @secure
         */
        this.linksDelete = (data, params = {}) => this.request(Object.assign({ path: `/links.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the requested link's information.
         *
         * @tags links
         * @name LinksGet
         * @summary Get Link
         * @request GET:/links.get
         * @secure
         */
        this.linksGet = (query, params = {}) => this.request(Object.assign({ path: `/links.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the requested link's information.
         *
         * @tags links
         * @name LinksGetPost
         * @summary Get Link (POST)
         * @request POST:/links.get
         * @secure
         */
        this.linksGetPost = (data, params = {}) => this.request(Object.assign({ path: `/links.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists the available links.
         *
         * @tags links
         * @name LinksList
         * @summary List Links
         * @request GET:/links.list
         * @secure
         */
        this.linksList = (query, params = {}) => this.request(Object.assign({ path: `/links.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists the available links.
         *
         * @tags links
         * @name LinksListPost
         * @summary List Links (POST)
         * @request POST:/links.list
         * @secure
         */
        this.linksListPost = (data, params = {}) => this.request(Object.assign({ path: `/links.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists metric definitions matching a filter.
         *
         * @tags slas
         * @name MetricDefinitionsList
         * @summary List Metric Definitions
         * @request GET:/metric-definitions.list
         * @secure
         */
        this.metricDefinitionsList = (query, params = {}) => this.request(Object.assign({ path: `/metric-definitions.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists metric definitions matching a filter.
         *
         * @tags slas
         * @name MetricDefinitionsListPost
         * @summary List Metric Definitions (POST)
         * @request POST:/metric-definitions.list
         * @secure
         */
        this.metricDefinitionsListPost = (data, params = {}) => this.request(Object.assign({ path: `/metric-definitions.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates new [part](https://devrev.ai/docs/product/parts).
         *
         * @tags parts
         * @name PartsCreate
         * @summary Create Part
         * @request POST:/parts.create
         * @secure
         */
        this.partsCreate = (data, params = {}) => this.request(Object.assign({ path: `/parts.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes a [part](https://devrev.ai/docs/product/parts).
         *
         * @tags parts
         * @name PartsDelete
         * @summary Delete Part
         * @request POST:/parts.delete
         * @secure
         */
        this.partsDelete = (data, params = {}) => this.request(Object.assign({ path: `/parts.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets a [part's](https://devrev.ai/docs/product/parts) information.
         *
         * @tags parts
         * @name PartsGet
         * @summary Get Part
         * @request GET:/parts.get
         * @secure
         */
        this.partsGet = (query, params = {}) => this.request(Object.assign({ path: `/parts.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets a [part's](https://devrev.ai/docs/product/parts) information.
         *
         * @tags parts
         * @name PartsGetPost
         * @summary Get Part (POST)
         * @request POST:/parts.get
         * @secure
         */
        this.partsGetPost = (data, params = {}) => this.request(Object.assign({ path: `/parts.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists a collection of [parts](https://devrev.ai/docs/product/parts).
         *
         * @tags parts
         * @name PartsList
         * @summary List Parts
         * @request GET:/parts.list
         * @secure
         */
        this.partsList = (query, params = {}) => this.request(Object.assign({ path: `/parts.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists a collection of [parts](https://devrev.ai/docs/product/parts).
         *
         * @tags parts
         * @name PartsListPost
         * @summary List Parts (POST)
         * @request POST:/parts.list
         * @secure
         */
        this.partsListPost = (data, params = {}) => this.request(Object.assign({ path: `/parts.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates a [part's](https://devrev.ai/docs/product/parts) information.
         *
         * @tags parts
         * @name PartsUpdate
         * @summary Update Part
         * @request POST:/parts.update
         * @secure
         */
        this.partsUpdate = (data, params = {}) => this.request(Object.assign({ path: `/parts.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates a Rev organization in the authenticated user's Dev organization.
         *
         * @tags rev-orgs
         * @name RevOrgsCreate
         * @summary Create Rev Org
         * @request POST:/rev-orgs.create
         * @secure
         */
        this.revOrgsCreate = (data, params = {}) => this.request(Object.assign({ path: `/rev-orgs.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes the Rev organization.
         *
         * @tags rev-orgs
         * @name RevOrgsDelete
         * @summary Delete Rev Org
         * @request POST:/rev-orgs.delete
         * @secure
         */
        this.revOrgsDelete = (data, params = {}) => this.request(Object.assign({ path: `/rev-orgs.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Retrieves the Rev organization's information.
         *
         * @tags rev-orgs
         * @name RevOrgsGet
         * @summary Get Rev Org
         * @request GET:/rev-orgs.get
         * @secure
         */
        this.revOrgsGet = (query, params = {}) => this.request(Object.assign({ path: `/rev-orgs.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Retrieves the Rev organization's information.
         *
         * @tags rev-orgs
         * @name RevOrgsGetPost
         * @summary Get Rev Org (POST)
         * @request POST:/rev-orgs.get
         * @secure
         */
        this.revOrgsGetPost = (data, params = {}) => this.request(Object.assign({ path: `/rev-orgs.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the list of Rev organizations' information belonging to the authenticated user's Dev Organization which the user is also authorized to access.
         *
         * @tags rev-orgs
         * @name RevOrgsList
         * @summary List Rev Orgs
         * @request GET:/rev-orgs.list
         * @secure
         */
        this.revOrgsList = (query, params = {}) => this.request(Object.assign({ path: `/rev-orgs.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the list of Rev organizations' information belonging to the authenticated user's Dev Organization which the user is also authorized to access.
         *
         * @tags rev-orgs
         * @name RevOrgsListPost
         * @summary List Rev Orgs (POST)
         * @request POST:/rev-orgs.list
         * @secure
         */
        this.revOrgsListPost = (data, params = {}) => this.request(Object.assign({ path: `/rev-orgs.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates the Rev organization's information.
         *
         * @tags rev-orgs
         * @name RevOrgsUpdate
         * @summary Update Rev Org
         * @request POST:/rev-orgs.update
         * @secure
         */
        this.revOrgsUpdate = (data, params = {}) => this.request(Object.assign({ path: `/rev-orgs.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates a Rev user for a Rev organization. Rev user can be a customer or a lead of an organization.
         *
         * @tags rev-users
         * @name RevUsersCreate
         * @summary Create Rev User
         * @request POST:/rev-users.create
         * @secure
         */
        this.revUsersCreate = (data, params = {}) => this.request(Object.assign({ path: `/rev-users.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes a Rev user.
         *
         * @tags rev-users
         * @name RevUsersDelete
         * @summary Delete Rev User
         * @request POST:/rev-users.delete
         * @secure
         */
        this.revUsersDelete = (data, params = {}) => this.request(Object.assign({ path: `/rev-users.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Returns the Rev user of a Rev organization by its ID.
         *
         * @tags rev-users
         * @name RevUsersGet
         * @summary Get Rev User
         * @request GET:/rev-users.get
         * @secure
         */
        this.revUsersGet = (query, params = {}) => this.request(Object.assign({ path: `/rev-users.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Returns the Rev user of a Rev organization by its ID.
         *
         * @tags rev-users
         * @name RevUsersGetPost
         * @summary Get Rev User (POST)
         * @request POST:/rev-users.get
         * @secure
         */
        this.revUsersGetPost = (data, params = {}) => this.request(Object.assign({ path: `/rev-users.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Returns a list of all Rev Users belonging to the authenticated user's Dev organization.
         *
         * @tags rev-users
         * @name RevUsersList
         * @summary List Rev Users
         * @request GET:/rev-users.list
         * @secure
         */
        this.revUsersList = (query, params = {}) => this.request(Object.assign({ path: `/rev-users.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Returns a list of all Rev Users belonging to the authenticated user's Dev organization.
         *
         * @tags rev-users
         * @name RevUsersListPost
         * @summary List Rev Users (POST)
         * @request POST:/rev-users.list
         * @secure
         */
        this.revUsersListPost = (data, params = {}) => this.request(Object.assign({ path: `/rev-users.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates a Rev user.
         *
         * @tags rev-users
         * @name RevUsersUpdate
         * @summary Update Rev User
         * @request POST:/rev-users.update
         * @secure
         */
        this.revUsersUpdate = (data, params = {}) => this.request(Object.assign({ path: `/rev-users.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets a service account.
         *
         * @tags service-accounts
         * @name ServiceAccountsGet
         * @summary Get Service Account
         * @request GET:/service-accounts.get
         * @secure
         */
        this.serviceAccountsGet = (query, params = {}) => this.request(Object.assign({ path: `/service-accounts.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets a service account.
         *
         * @tags service-accounts
         * @name ServiceAccountsGetPost
         * @summary Get Service Account (POST)
         * @request POST:/service-accounts.get
         * @secure
         */
        this.serviceAccountsGetPost = (data, params = {}) => this.request(Object.assign({ path: `/service-accounts.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets an SLA tracker.
         *
         * @tags slas
         * @name SlaTrackersGet
         * @summary Get Sla Tracker
         * @request GET:/sla-trackers.get
         * @secure
         */
        this.slaTrackersGet = (query, params = {}) => this.request(Object.assign({ path: `/sla-trackers.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets an SLA tracker.
         *
         * @tags slas
         * @name SlaTrackersGetPost
         * @summary Get Sla Tracker (POST)
         * @request POST:/sla-trackers.get
         * @secure
         */
        this.slaTrackersGetPost = (data, params = {}) => this.request(Object.assign({ path: `/sla-trackers.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists SLA trackers matching a filter.
         *
         * @tags slas
         * @name SlaTrackersList
         * @summary List Sla Trackers
         * @request GET:/sla-trackers.list
         * @secure
         */
        this.slaTrackersList = (query, params = {}) => this.request(Object.assign({ path: `/sla-trackers.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists SLA trackers matching a filter.
         *
         * @tags slas
         * @name SlaTrackersListPost
         * @summary List Sla Trackers (POST)
         * @request POST:/sla-trackers.list
         * @secure
         */
        this.slaTrackersListPost = (data, params = {}) => this.request(Object.assign({ path: `/sla-trackers.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description  Assigns the SLA to a set of Rev organizations.
         *
         * @tags slas
         * @name SlasAssign
         * @summary Assign Slas
         * @request POST:/slas.assign
         * @secure
         */
        this.slasAssign = (data, params = {}) => this.request(Object.assign({ path: `/slas.assign`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates an SLA in draft status.
         *
         * @tags slas
         * @name SlasCreate
         * @summary Create Sla
         * @request POST:/slas.create
         * @secure
         */
        this.slasCreate = (data, params = {}) => this.request(Object.assign({ path: `/slas.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets an SLA.
         *
         * @tags slas
         * @name SlasGet
         * @summary Get Sla
         * @request GET:/slas.get
         * @secure
         */
        this.slasGet = (query, params = {}) => this.request(Object.assign({ path: `/slas.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets an SLA.
         *
         * @tags slas
         * @name SlasGetPost
         * @summary Get Sla (POST)
         * @request POST:/slas.get
         * @secure
         */
        this.slasGetPost = (data, params = {}) => this.request(Object.assign({ path: `/slas.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists SLAs matching a filter.
         *
         * @tags slas
         * @name SlasList
         * @summary List Slas
         * @request GET:/slas.list
         * @secure
         */
        this.slasList = (query, params = {}) => this.request(Object.assign({ path: `/slas.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists SLAs matching a filter.
         *
         * @tags slas
         * @name SlasListPost
         * @summary List Slas (POST)
         * @request POST:/slas.list
         * @secure
         */
        this.slasListPost = (data, params = {}) => this.request(Object.assign({ path: `/slas.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Changes the status of an SLA.
         *
         * @tags slas
         * @name SlasTransition
         * @summary Transition Slas
         * @request POST:/slas.transition
         * @secure
         */
        this.slasTransition = (data, params = {}) => this.request(Object.assign({ path: `/slas.transition`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates a draft SLA.
         *
         * @tags slas
         * @name SlasUpdate
         * @summary Update Sla
         * @request POST:/slas.update
         * @secure
         */
        this.slasUpdate = (data, params = {}) => this.request(Object.assign({ path: `/slas.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists system users within your organization.
         *
         * @tags sys-users
         * @name SysUsersList
         * @summary List Sys Users
         * @request GET:/sys-users.list
         * @secure
         */
        this.sysUsersList = (query, params = {}) => this.request(Object.assign({ path: `/sys-users.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists system users within your organization.
         *
         * @tags sys-users
         * @name SysUsersListPost
         * @summary List Sys Users (POST)
         * @request POST:/sys-users.list
         * @secure
         */
        this.sysUsersListPost = (data, params = {}) => this.request(Object.assign({ path: `/sys-users.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates the system user.
         *
         * @tags sys-users
         * @name SysUsersUpdate
         * @summary Update Sys User
         * @request POST:/sys-users.update
         * @secure
         */
        this.sysUsersUpdate = (data, params = {}) => this.request(Object.assign({ path: `/sys-users.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates a new tag, which is used to create associations between objects and a logical concept denoted by the tag's name.
         *
         * @tags tags
         * @name TagsCreate
         * @summary Create Tag
         * @request POST:/tags.create
         * @secure
         */
        this.tagsCreate = (data, params = {}) => this.request(Object.assign({ path: `/tags.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes a tag.
         *
         * @tags tags
         * @name TagsDelete
         * @summary Delete Tag
         * @request POST:/tags.delete
         * @secure
         */
        this.tagsDelete = (data, params = {}) => this.request(Object.assign({ path: `/tags.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets a tag's information.
         *
         * @tags tags
         * @name TagsGet
         * @summary Get Tag
         * @request GET:/tags.get
         * @secure
         */
        this.tagsGet = (query, params = {}) => this.request(Object.assign({ path: `/tags.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets a tag's information.
         *
         * @tags tags
         * @name TagsGetPost
         * @summary Get Tag (POST)
         * @request POST:/tags.get
         * @secure
         */
        this.tagsGetPost = (data, params = {}) => this.request(Object.assign({ path: `/tags.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists the available tags.
         *
         * @tags tags
         * @name TagsList
         * @summary List Tags
         * @request GET:/tags.list
         * @secure
         */
        this.tagsList = (query, params = {}) => this.request(Object.assign({ path: `/tags.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists the available tags.
         *
         * @tags tags
         * @name TagsListPost
         * @summary List Tags (POST)
         * @request POST:/tags.list
         * @secure
         */
        this.tagsListPost = (data, params = {}) => this.request(Object.assign({ path: `/tags.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates a tag's information.
         *
         * @tags tags
         * @name TagsUpdate
         * @summary Update Tag
         * @request POST:/tags.update
         * @secure
         */
        this.tagsUpdate = (data, params = {}) => this.request(Object.assign({ path: `/tags.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates a new entry on an object's timeline.
         *
         * @tags timeline-entries
         * @name TimelineEntriesCreate
         * @summary Create Timeline Entry
         * @request POST:/timeline-entries.create
         * @secure
         */
        this.timelineEntriesCreate = (data, params = {}) => this.request(Object.assign({ path: `/timeline-entries.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes an entry from an object's timeline.
         *
         * @tags timeline-entries
         * @name TimelineEntriesDelete
         * @summary Delete Timeline Entry
         * @request POST:/timeline-entries.delete
         * @secure
         */
        this.timelineEntriesDelete = (data, params = {}) => this.request(Object.assign({ path: `/timeline-entries.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets an entry on an object's timeline.
         *
         * @tags timeline-entries
         * @name TimelineEntriesGet
         * @summary Get Timeline Entry
         * @request GET:/timeline-entries.get
         * @secure
         */
        this.timelineEntriesGet = (query, params = {}) => this.request(Object.assign({ path: `/timeline-entries.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets an entry on an object's timeline.
         *
         * @tags timeline-entries
         * @name TimelineEntriesGetPost
         * @summary Get Timeline Entry (POST)
         * @request POST:/timeline-entries.get
         * @secure
         */
        this.timelineEntriesGetPost = (data, params = {}) => this.request(Object.assign({ path: `/timeline-entries.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists the timeline entries for an object.
         *
         * @tags timeline-entries
         * @name TimelineEntriesList
         * @summary List Timeline Entries
         * @request GET:/timeline-entries.list
         * @secure
         */
        this.timelineEntriesList = (query, params = {}) => this.request(Object.assign({ path: `/timeline-entries.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists the timeline entries for an object.
         *
         * @tags timeline-entries
         * @name TimelineEntriesListPost
         * @summary List Timeline Entries (POST)
         * @request POST:/timeline-entries.list
         * @secure
         */
        this.timelineEntriesListPost = (data, params = {}) => this.request(Object.assign({ path: `/timeline-entries.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates an entry on an object's timeline.
         *
         * @tags timeline-entries
         * @name TimelineEntriesUpdate
         * @summary Update Timeline Entry
         * @request POST:/timeline-entries.update
         * @secure
         */
        this.timelineEntriesUpdate = (data, params = {}) => this.request(Object.assign({ path: `/timeline-entries.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates a new webhook target.
         *
         * @tags webhooks
         * @name WebhooksCreate
         * @summary Create Webhook
         * @request POST:/webhooks.create
         * @secure
         */
        this.webhooksCreate = (data, params = {}) => this.request(Object.assign({ path: `/webhooks.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes the requested webhook.
         *
         * @tags webhooks
         * @name WebhooksDelete
         * @summary Delete Webhook
         * @request POST:/webhooks.delete
         * @secure
         */
        this.webhooksDelete = (data, params = {}) => this.request(Object.assign({ path: `/webhooks.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Describes a webhook invocation for an event from DevRev to a webhook's target URL, where the receiving handler must implement the specified protocol. Note the documented endpoint is for exposition and not provided by DevRev.
         *
         * @tags webhooks
         * @name WebhooksEvent
         * @summary Event Webhooks
         * @request POST:/webhooks.event
         * @secure
         */
        this.webhooksEvent = (data, params = {}) => this.request(Object.assign({ path: `/webhooks.event`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets the requested webhook's information.
         *
         * @tags webhooks
         * @name WebhooksGet
         * @summary Get Webhook
         * @request GET:/webhooks.get
         * @secure
         */
        this.webhooksGet = (query, params = {}) => this.request(Object.assign({ path: `/webhooks.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets the requested webhook's information.
         *
         * @tags webhooks
         * @name WebhooksGetPost
         * @summary Get Webhook (POST)
         * @request POST:/webhooks.get
         * @secure
         */
        this.webhooksGetPost = (data, params = {}) => this.request(Object.assign({ path: `/webhooks.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists the webhooks.
         *
         * @tags webhooks
         * @name WebhooksList
         * @summary List Webhooks
         * @request GET:/webhooks.list
         * @secure
         */
        this.webhooksList = (params = {}) => this.request(Object.assign({ path: `/webhooks.list`, method: 'GET', secure: true, format: 'json' }, params));
        /**
         * @description Lists the webhooks.
         *
         * @tags webhooks
         * @name WebhooksListPost
         * @summary List Webhooks (POST)
         * @request POST:/webhooks.list
         * @secure
         */
        this.webhooksListPost = (data, params = {}) => this.request(Object.assign({ path: `/webhooks.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates the requested webhook.
         *
         * @tags webhooks
         * @name WebhooksUpdate
         * @summary Update Webhook
         * @request POST:/webhooks.update
         * @secure
         */
        this.webhooksUpdate = (data, params = {}) => this.request(Object.assign({ path: `/webhooks.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Creates new work ([issue](https://devrev.ai/docs/product/build), [ticket](https://devrev.ai/docs/product/support)) item. [task](https://devrev.ai/docs/product/tasks) and opportunity work types are supported in the beta version.
         *
         * @tags works
         * @name WorksCreate
         * @summary Create Work
         * @request POST:/works.create
         * @secure
         */
        this.worksCreate = (data, params = {}) => this.request(Object.assign({ path: `/works.create`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Deletes a work item.
         *
         * @tags works
         * @name WorksDelete
         * @summary Delete Work
         * @request POST:/works.delete
         * @secure
         */
        this.worksDelete = (data, params = {}) => this.request(Object.assign({ path: `/works.delete`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Exports a collection of work items.
         *
         * @tags works
         * @name WorksExport
         * @summary Export Works
         * @request GET:/works.export
         * @secure
         */
        this.worksExport = (query, params = {}) => this.request(Object.assign({ path: `/works.export`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Exports a collection of work items.
         *
         * @tags works
         * @name WorksExportPost
         * @summary Export Works (POST)
         * @request POST:/works.export
         * @secure
         */
        this.worksExportPost = (data, params = {}) => this.request(Object.assign({ path: `/works.export`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Gets a work item's information.
         *
         * @tags works
         * @name WorksGet
         * @summary Get Work
         * @request GET:/works.get
         * @secure
         */
        this.worksGet = (query, params = {}) => this.request(Object.assign({ path: `/works.get`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Gets a work item's information.
         *
         * @tags works
         * @name WorksGetPost
         * @summary Get Work (POST)
         * @request POST:/works.get
         * @secure
         */
        this.worksGetPost = (data, params = {}) => this.request(Object.assign({ path: `/works.get`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Lists a collection of work items.
         *
         * @tags works
         * @name WorksList
         * @summary List Works
         * @request GET:/works.list
         * @secure
         */
        this.worksList = (query, params = {}) => this.request(Object.assign({ path: `/works.list`, method: 'GET', query: query, secure: true, format: 'json' }, params));
        /**
         * @description Lists a collection of work items.
         *
         * @tags works
         * @name WorksListPost
         * @summary List Works (POST)
         * @request POST:/works.list
         * @secure
         */
        this.worksListPost = (data, params = {}) => this.request(Object.assign({ path: `/works.list`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
        /**
         * @description Updates a work item's information.
         *
         * @tags works
         * @name WorksUpdate
         * @summary Update Work
         * @request POST:/works.update
         * @secure
         */
        this.worksUpdate = (data, params = {}) => this.request(Object.assign({ path: `/works.update`, method: 'POST', body: data, secure: true, type: ContentType.Json, format: 'json' }, params));
    }
}
exports.Api = Api;
